nilearn/region_extraction/regionextractor.py

```diff
@@ -0,0 +1,341 @@
+                 target_shape=None, standardize=False, low_pass=None,
+                 high_pass=None, t_r=None, memory=Memory(cachedir=None),
+                 min_size=20, threshold_strategy='voxelratio', threshold='auto',
+                 extractor='voxel_wise', smooth_fwhm=6., verbose=0):
+        self.n_regions = n_regions
+        self.mask = mask
+        self.target_affine = target_affine
+        self.target_shape = target_shape
+        self.standardize = standardize
+        self.low_pass = low_pass
+        self.high_pass = high_pass
+        self.t_r = t_r
+        self.memory = memory
+
+        self.min_size = min_size
+        self.threshold_strategy = threshold_strategy
+        self.threshold = threshold
+        self.extractor = extractor
+        self.smooth_fwhm = smooth_fwhm
+        self.verbose = verbose
+
+    def fit(self, maps_img):
+        """ Compute the mask and fit the mask to the maps data and
+        and extract or seperate the regions into region from the maps.
+
+        Parameters
+        ----------
+        maps_img: a Niimg-like image/object.
+            the image which consists of atlas maps or statistically
+            estimated maps.
+
+        Returns
+        -------
+        regions_: 4D Niimg-like image/objects.
+            a list of seperated region converted to Niimg concatenated
+            in a 4th dimension.
+
+        index_: a numpy array
+            a list of indices in an array to see the identity of each
+            seperated region to a family of atlas maps.
+        """
+        maps_img = check_niimg(maps_img)
+        len_maps = maps_img.shape[:3]
+        maps_data = maps_img.get_data()
+        self.maps_data = maps_data
+        affine = maps_img.get_affine()
+        min_size = self.min_size
+        if isinstance(self.mask, (_basestring, nibabel.Nifti1Image)):
+            mask = self.mask
+            mask_img = check_niimg_3d(mask)
+            if maps_img.shape[:3] != mask_img.shape:
+                raise ValueError("Provided mask image 'mask' and "
+                                 "3D shape of input maps must have "
+                                 "identical shape")
+            else:
+                self.masker_.mask_img_ = mask
+        elif isinstance(self.mask, NiftiMasker):
+            self.masker_ = clone(self.mask)
+        elif self.mask is None:
+            self.masker_ = NiftiMasker(mask_img=self.mask,
+                                       target_affine=self.target_affine,
+                                       target_shape=self.target_shape,
+                                       standardize=self.standardize,
+                                       low_pass=self.low_pass,
+                                       high_pass=self.high_pass,
+                                       mask_strategy='background',
+                                       t_r=self.t_r,
+                                       memory=self.memory)
+
```

Comment:
Your whole block is good but it can be factorized. It's better since it's shorter and you rely on nilearn's code for handling exceptions.
```python
# First create the masker
if isinstance(self.mask, NiftiMasker):
    self.masker_ = clone(self.mask)
else:
    self.masker_ = NiftiMasker(mask_img=self.mask, ...)

# Then fit
if self.masker_.mask_img is None:
    self.masker_.fit(maps_img)
else:
    self.masker_.fit()
self.mask_img_ = self.masker_.mask_img_